source/extensions/clusters/common/logical_host.cc

```diff
@@ -6,14 +6,14 @@ namespace Upstream {
 Upstream::Host::CreateConnectionData LogicalHost::createConnection(
     Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
     Network::TransportSocketOptionsConstSharedPtr transport_socket_options) const {
-  const auto current_address = address();
-  const std::vector<Network::Address::InstanceConstSharedPtr>& address_list = addressList();
-
+  const std::pair<Network::Address::InstanceConstSharedPtr,
+                  const std::vector<Network::Address::InstanceConstSharedPtr>>
+      address = addressAndListCopy();
```

Comment:
I've suggested it IRL, but just adding it here:
The vector copy is being done on each connection creation. An alternative could be changing the type of the vector to:
`std::shared_ptr<const vector<...>>`. This way it will only be copied once, and shared between all its users.

Reply:
+1

Reply:
Thanks for the comments!

Currently the copy is done for one pointer under the read lock, and this PR extended that to a couple of address pointers. I would think the performance impact is negligible. But, we can run a load test to prove that if that's necessary.

BTW, if we change the address_list to type std::shared_ptr<const vector<...>>, it changes the data structure of the base class which will have impact all the host types.   As there is no evidence that the issue exist for non-logical-host types, I would consider keep the change inside logical_host is safer. WDYT?   Also, I am open to the idea to have a cleanup PR to explore in this direction.

Reply:
Per whiteboard discusison and chat, Yanjun and I discussed this idea and want to get a  taste-test from @adisuissa and @yanavlasov 

The idea is to leave HostImpl alone: this is not something we want to touch as it additional locking or complexity may impact its performance in the most common use case.

We want the address-list in LogicalHost to be different in 2 ways:
 * absl lock annotations
 * `shared_ptr<const vector<>>`

So I propose we just add those member vars to LogicalHost and not use the ones in the parent class; they'd just remain empty. As an option, we could cut that inheritance if there's nothing we really use from HostImpl. I haven't looked to see if that's feasible or if it is functionally correct to leave HostImpl's member vars unpopulated.

Reply:
This PR implemented the idea to use a shared_ptr to store the address_list: https://github.com/envoyproxy/envoy/pull/32830/.

This avoided the copy the entire vector during READ operations.

Reply:
For your comments on add " address_ and address_list" into LogicalHost, and keep the parent class ones empty, we now know the parent ones can not be empty.   They have to be setup for other features like heath checking: https://github.com/envoyproxy/envoy/blob/10a10039b3a82d43ff47c319e0ef4faf229f3327/source/common/upstream/upstream_impl.cc#L413, 
 crypto: https://github.com/envoyproxy/envoy/blob/10a10039b3a82d43ff47c319e0ef4faf229f3327/source/common/upstream/upstream_impl.cc#L421, ....

If we can not keep the base class ones empty, then I am hesitating to adding a copy in the derived class.  Once we have it in both places, we have to keep them in sync, like updating both of them at once under lock, ....  If we use a shared_ptr at both places, then we need to make sure the lifetime are maintained correctly.

So, please take a look https://github.com/envoyproxy/envoy/pull/32830.  I think it is a good solution, and no data model change at all.